src/CloudNative.CloudEvents.Amqp/AmqpCloudEventMessage.cs

```diff
@@ -1,90 +1,96 @@
         {
+            ApplicationProperties = new ApplicationProperties();
+            MapHeaders(cloudEvent);
+
             if (contentMode == ContentMode.Structured)
             {
-                this.BodySection = new Data
-                    { Binary = formatter.EncodeStructuredEvent(cloudEvent, out var contentType) };
-                this.Properties = new Properties() { ContentType = contentType.MediaType };
-                this.ApplicationProperties = new ApplicationProperties();
+                BodySection = new Data
+                {
+                    Binary = formatter.EncodeStructuredEvent(cloudEvent, out var contentType)
+                };
+                Properties = new Properties { ContentType = contentType.MediaType };
+                ApplicationProperties = new ApplicationProperties();
                 MapHeaders(cloudEvent);
                 return;
             }
-
-            if (cloudEvent.Data is byte[])
+            else
             {
-                this.BodySection = new Data { Binary = (byte[])cloudEvent.Data };
+                BodySection = SerializeData(cloudEvent.Data);
+                Properties = new Properties { ContentType = cloudEvent.DataContentType };
             }
-            else if (cloudEvent.Data is Stream)
+        }
+
+        private void MapHeaders(CloudEvent cloudEvent)
+        {
+            var properties = ApplicationProperties.Map;
+            properties.Add(AmqpClientExtensions.SpecVersionAmqpHeader, cloudEvent.SpecVersion.VersionId);
+            
+            foreach (var pair in cloudEvent.GetPopulatedAttributes())
             {
-                if (cloudEvent.Data is MemoryStream)
+                var attribute = pair.Key;
+
+                // The content type is specified elsewhere.
+                if (attribute == cloudEvent.SpecVersion.DataContentTypeAttribute)
                 {
-                    this.BodySection = new Data { Binary = ((MemoryStream)cloudEvent.Data).ToArray() };
+                    continue;
                 }
-                else
+
+                string propKey = AmqpClientExtensions.AmqpHeaderPrefix + attribute.Name;
+
+                // TODO: Check that AMQP can handle byte[], bool and int values
+                object propValue = pair.Value switch
                 {
-                    var buffer = new MemoryStream();
-                    ((Stream)cloudEvent.Data).CopyTo(buffer);
-                    this.BodySection = new Data { Binary = buffer.ToArray() };
-                }
+                    Uri uri => uri.ToString(),
+                    // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
+                    // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
+                    DateTimeOffset dto => dto.UtcDateTime,
+                    _ => pair.Value
```

Comment:
It looks like previously the default case would result in inserting a new `Map` as the value in the outer `Map` with all of the non-uri/DTO values. Is this no longer equivalent?

Reply:
Hmm... I think you're right, but I suspect that that was *either* unintentional, *or* it was a hangover from when CloudEvent attributes could contain maps in JSON format. I don't think we want it, either way. (CloudEvent attributes are very much single-valued now.)